### PR TITLE
yami_vpp: colorbalance setting

### PR DESCRIPTION
**Pull request by wudping**
implement colorbalance setting for hue, saturation, brightness and contrast.

Signed-off-by: wudping <dongpingx.wu@intel.com>

### CODE REVIEW

**Comment by wudping** [comment]
corresponding libyami-utils as following:
https://github.com/01org/libyami-utils/pull/57

File: vpp/vaapipostprocess_scaler.cpp [hunk 1]
```diff
@@ -135,6 +147,7 @@ bool VaapiPostProcessScaler::mapToRange(
         return false;
 
     value = min + (max - min) / (maxLevel - minLevel) * level;
+
```
Comment (chivakker) [inline vpp/vaapipostprocess_scaler.cpp:150]: unnecessary change.

**Review by chivakker — COMMENTED** [review]
This change looks good to me other that the unnecessary change.

File: vpp/vaapipostprocess_scaler.cpp (hunk 1 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp:150]: thanks @chivakker , I will delete line 150.

File: vpp/vaapipostprocess_scaler.cpp [hunk 2]
```diff
@@ -219,6 +232,72 @@ VaapiPostProcessScaler::createDeinterlaceFilter(const VPPDeinterlaceParameters&
...
+    }
+    return COLORBALANCE_NONE;
+}
+
+YamiStatus
+VaapiPostProcessScaler::createColorBalanceFilter(ColorBalanceParams* colorbalanceparam, const VPPColorBalanceParameters& colorbalance, const VAProcFilterValueRange& range)
+{
+    float value;
+
+    mapToRange(value, range.min_value, range.max_value, colorbalance.level, COLORBALANCE_LEVEL_MIN, COLORBALANCE_LEVEL_MAX);
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: please check mapToRange return value

File: vpp/vaapipostprocess_scaler.cpp [hunk 3]
```diff
@@ -219,6 +232,72 @@ VaapiPostProcessScaler::createDeinterlaceFilter(const VPPDeinterlaceParameters&
...
+
+    VAProcFilterParameterBufferColorBalance* d;
+    colorbalanceparam->filter = VaapiBuffer::create(m_context, VAProcFilterParameterBufferType, d);
+    if (!colorbalanceparam->filter)
+        return YAMI_DRIVER_FAIL;
+    d->type = VAProcFilterColorBalance;
+    d->attrib = mapToVaColorBalanceMode(colorbalance.mode);
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: Please check the return value

File: vpp/vaapipostprocess_scaler.cpp [hunk 4]
```diff
@@ -219,6 +232,72 @@ VaapiPostProcessScaler::createDeinterlaceFilter(const VPPDeinterlaceParameters&
     return YAMI_SUCCESS;
 }
 
+static VAProcColorBalanceType mapToVaColorBalanceMode(VppColorBalanceMode mode)
+{
+    switch (mode) {
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: so many switch cases, every error prone, could you help to change to Array loop style?
Just like https://github.com/01org/libyami/blob/ef7953bc7196c959051a368b5f2c0a71ad4a0a37/v4l2/v4l2_codecbase.cpp#L620
and  https://github.com/01org/libyami/blob/ef7953bc7196c959051a368b5f2c0a71ad4a0a37/v4l2/v4l2_codecbase.cpp#L633

File: vpp/vaapipostprocess_scaler.cpp [hunk 5]
```diff
@@ -249,6 +328,61 @@ YamiStatus VaapiPostProcessScaler::setDeinterlaceParam(const VPPDeinterlaceParam
...
+
+    VAProcFilterCapColorBalance caps[VAProcColorBalanceCount];
+    if (supportedModeRanges.empty()) {
+        //query from libva.
+        uint32_t num = VAProcColorBalanceCount;
+        YamiStatus status = queryVideoProcFilterCaps(VAProcFilterColorBalance, caps, &num);
+        if (status != YAMI_SUCCESS)
+            return status;
+        for (uint32_t i = 0; i < num; i++) {
+            modeRange.mode = mapToVppColorBalanceMode(caps[i].type);
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: Please check and drop unsupported mode.

File: vpp/vaapipostprocess_scaler.cpp [hunk 6]
```diff
@@ -249,6 +328,61 @@ YamiStatus VaapiPostProcessScaler::setDeinterlaceParam(const VPPDeinterlaceParam
...
+            modeRange.range = caps[i].range;
+            supportedModeRanges.insert(modeRange);
+        }
+    }
+    if (colorbalance.mode == COLORBALANCE_NONE) {
+        m_brightness.filter.reset();
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: this wired, you treat COLORBALANCE_NONE as brightness? what about other filters?

File: vpp/vaapipostprocess_scaler.cpp [hunk 7]
```diff
@@ -249,6 +328,61 @@ YamiStatus VaapiPostProcessScaler::setDeinterlaceParam(const VPPDeinterlaceParam
...
+        return YAMI_SUCCESS;
+    }
+    ColorBalanceItr iteratorMR;
+    ColorBalanceModeRange clMR;
+    clMR.mode = colorbalance.mode;
+    iteratorMR = supportedModeRanges.find(clMR);
+    if (iteratorMR == supportedModeRanges.end()) {
+        m_brightness.filter.reset();
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: you treat none as brightness again

File: vpp/vaapipostprocess_scaler.h [hunk 8]
```diff
@@ -48,6 +48,26 @@ class VaapiPostProcessScaler : public VaapiPostProcessBase {
...
+            return (mr1.mode < mr2.mode) ? true : false;
+        }
+    };
+
+    typedef std::set<ColorBalanceModeRange, ColorBalanceCompare> ColorBalanceList;
+    typedef std::set<ColorBalanceModeRange, ColorBalanceCompare>::iterator ColorBalanceItr;
+    struct ColorBalanceParams {
+        VppColorBalanceMode mode; //set by user
+        ColorBalanceList supportedModeRanges; //query from va
+        //std::set<ColorBalanceModeRange, ColorBalanceModeRangeCompare> supportedModeRanges;
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.h]: remove this

File: vpp/vaapipostprocess_scaler.h [hunk 9]
```diff
@@ -48,6 +48,26 @@ class VaapiPostProcessScaler : public VaapiPostProcessBase {
...
+        bool operator()(const ColorBalanceModeRange& mr1, const ColorBalanceModeRange& mr2) const
+        {
+            return (mr1.mode < mr2.mode) ? true : false;
+        }
+    };
+
+    typedef std::set<ColorBalanceModeRange, ColorBalanceCompare> ColorBalanceList;
+    typedef std::set<ColorBalanceModeRange, ColorBalanceCompare>::iterator ColorBalanceItr;
+    struct ColorBalanceParams {
+        VppColorBalanceMode mode; //set by user
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.h]: why we need mode here and  L52

**Review by xuguangxin — CHANGES_REQUESTED** [review]

File: vpp/vaapipostprocess_scaler.cpp (hunk 6 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks @xuguangxin for your constructive suggestion.
How about to reset all the colorbalance filters, such as hue, contrast, brightness and saturation,  when getting COLORBALANCE_NONE ?

Comment (wudping) [inline vpp/vaapipostprocess_scaler.cpp]: thanks @xuguangxin .

File: vpp/vaapipostprocess_scaler.cpp (hunk 7 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks @xuguangxin .

File: vpp/vaapipostprocess_scaler.cpp (hunk 4 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks @xuguangxin for advises.

File: vpp/vaapipostprocess_scaler.cpp (hunk 2 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks for suggestion @xuguangxin .

File: vpp/vaapipostprocess_scaler.cpp (hunk 3 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks for suggestion @xuguangxin .

File: vpp/vaapipostprocess_scaler.cpp (hunk 5 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.cpp]: thanks @xuguangxin .

File: vpp/vaapipostprocess_scaler.h (hunk 9 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.h]: I just put VppColorBalanceMode and VAProcFilterValueRange together, so I can find out the range value via VppColorBalanceMode. 
But I will save them in a Map, according to your suggestion.
Thanks @xuguangxin .

File: vpp/vaapipostprocess_scaler.h (hunk 8 above)
Comment (wudping) [reply vpp/vaapipostprocess_scaler.h]: okay ,thanks.

**Comment by wudping** [comment]
updated according to the detail comments;
please help to review.

File: vpp/vaapipostprocess_scaler.cpp [hunk 10]
```diff
@@ -249,6 +339,38 @@ YamiStatus VaapiPostProcessScaler::setDeinterlaceParam(const VPPDeinterlaceParam
     return createDeinterlaceFilter(deinterlace);
 }
 
+YamiStatus VaapiPostProcessScaler::setColorBalanceParam(const VPPColorBalanceParameters& colorbalance)
+{
+    VAProcFilterCapColorBalance caps[VAProcColorBalanceCount];
+    static bool rangeInit = false;
```
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp]: Please do not use static, we may support difference va at same time, ie card0 or card1.

**Review by xuguangxin — APPROVED** [review]

**Comment by xuguangxin** [comment]
verify +1

**Comment by xuguangxin** [comment]
merged as 30c7aa00b3a27b4c8802f7a755c599ccf5677a11

**Pull request closed without merging**